### ISSUE TITLE
Standardize on extern modifier instead of empty Method bodies

### ISSUE THREAD

**Issue by TinkerWorX** [bridgedotnet/Bridge#828]
Is there a reason the `extern` semantic isn't used with `Template`? I took a look at DateTime.cs, and saw a lot of empty or meaningless bodies for pretty much all functions. I have not yet looked into the toolchain involved in making C# into JS, but I have experience using `Mono.Cecil.AssemblyDefinition` to attach a body to methods post-compile, which could be an option for creating a cleaner and clearer codebase.

**Comment by geoffreymcgill** [comment]
Hi @TinkerWorX, 

Those empty Method bodies are no longer required and the `extern` modifier can be applied to the Method instead. When those definition files were initially created we were not yet using the `extern` modifier. 

**Example**

```csharp
// Existing
public int GetFullYear()
{
    return 0;
}

// Revised
public extern int GetFullYear()
```

We will clean up those definition files, although at the moment that's a bit of a low priority. Nice project for the community. ;)

**Comment by geoffreymcgill** [comment]
I revised the title of this Issue and will use to track our progress with standardizing on the `extern` modifier. I will have to change the Milestone to 1.12 though.

**Comment by TinkerWorX** [comment]
That's good to hear, just wanted to make sure. I'll keep in mind that it's ready to be updated and I'll see if I can throw a few PRs when I get the time. :)

**Comment by AndreyChechel** [comment]
> Nice project for the community. ;)

Challenge accepted =)

I've created a tool that can help you automate this monotonous work. It enumerates all files in the specified directories and updates method signatures using Roslyn. This utility was tested for the Bridge project and I was able to compile the project after all method signatures had been updated. 

On a less positive note - Bridge.Html5 project. The methods were updated successfully, however the project couldn't be compiled - a batch of errors connected with DllImport attribute appeared. I'd like to take a look and try to find out what are the differences between those projects, will keep you informed. If you know the reason, please drop me a line, it'd be greatly appreciated.

Please feel free to get sources and use the Utility:
https://github.com/AndreyChechel/Bridge-ExternUtility

**Comment by geoffreymcgill** [comment]
@AndreyChechel: Very very cool. Thanks for taking this on. Didn't expect that quick of a turn around on the request. We'll give your utility a try. 

I'll get @vladsch to consult on the **Bridge.Html5** issues you noted.

**Comment by TinkerWorX** [comment]
@AndreyChechel I can also help out by checking the output if you can show me how to do the conversion or if you can share the output.

**Comment by geoffreymcgill** [comment]
Hi @AndreyChechel: Looks like you just revise the directory path to your locally cloned Bridge folder, then run the Console app. 

https://github.com/AndreyChechel/Bridge-ExternUtility/blob/master/BridgeExternUtility/BridgeExternUtility/Program.cs#L19

I haven't had a chance to run the utility yet, so maybe there's more required, but it looks like that's it.

**Comment by AndreyChechel** [comment]
Hi @geoffreymcgill, thanks for a quick answer and for you willing to help!

As for you comment: right, I'm processing my local Bridge folder. In order to run the utility for your projects you should replace the path with your local directory. You can also uncomment and update the path to Brdige.Html5 project (I commented it until the issue with that project is fixed).

If you want to review the way the utility works, please take a look to the RoslynHelper class:
https://github.com/AndreyChechel/Bridge-ExternUtility/blob/master/BridgeExternUtility/BridgeExternUtility/RoslynHelper.cs

**Comment by AndreyChechel** [comment]
Hi @TinkerWorX, thanks - it'd be great if you had a chance to check the utility! You should download sources from the referenced repository. Replace paths for Bridge / Bridge.Html5 projects to your local ones (in Program.cs (lines 19-20)). That's it, now you can build and run this console app.

**Comment by TinkerWorX** [comment]
@AndreyChechel, I found the problem. You were updating virtual methods. I made a PR on your project with a fix.

**Comment by AndreyChechel** [comment]
Hi @TinkerWorX, thank you for the suggestion! Well, it seems virtual methods can work with extern modifier. You can check it by looking at `Date.cs` in `Bridge` project. This class contains a set of virtual methods (for instance, `ToDateString`), nevertheless the project could be compiled.

I've spent some time and discovered that there is no magic here and the reason why `Bridge.Html5` couldn't be compiled is Warning 0626 which is suppressed in `Bridge` but is not suppressed in `Bridge.Html5`. After I suppressed the warning in `Bridge.Html5` I was able to run the Utility for the both projects without issues.

**Comment by geoffreymcgill** [comment]
> Warning 0626 which is suppressed in Bridge but is not suppressed in Bridge.Html5

We will fix this issue. Thanks for pointing us in the right direction.

I was just about to try out your utility, so I'll let you know if I run into any issues.

**Comment by TinkerWorX** [comment]
Great to hear you got it working. How peculiar that a warning would cause a compiler error like that. My fix was based on the simple fact that `virtual extern` seemed like a thing that shouldn't be possible, but come to think of it, I guess you could override the base method, even if it's `extern`.

**Comment by geoffreymcgill** [comment]
I fixed the **DllImport** warning issue, see #838. 

The fix will be sent as a pull request right away, then integrated into **master** branch on Friday, Jan 22.

**Comment by geoffreymcgill** [comment]
Hi @AndreyChechel,

I tested **Bridge-ExternUtility** and it seems to work really well. I can update **Bridge** and **Bridge.Html5** repos with the revised `extern` modifiers. Both projects compile properly. 

Can you add a [LICENSE](https://github.com/bridgedotnet/Bridge/blob/master/LICENSE) document to your project? 

Once you choose a license, preferably Apache 2.0, we should be able to officially use the utility to modify the projects. It shouldn't be a problem to include these changes in the upcoming 1.11 release. I have updated the Milestone to 1.11.

**Comment by AndreyChechel** [comment]
Hi @geoffreymcgill, glad to hear that the utility works without issues. I've added Apache 2.0 license to the project as you requested.
Besides that I've implemented another feature - processing of `operator` overloadings with empty bodies. I was able to found a few occurrences in **Bridge** for that case.

**Comment by geoffreymcgill** [comment]
Hi @AndreyChechel: Thanks for the update and for applying the LICENSE. 

Today we are working on getting all the branches and pull-requests merged into the **master** branch. Once that is complete, I will run your **ExternUtility** on **Bridge** and **Html5** projects.
